### DISCUSSION TITLE
Syntactic sugar for primary constructors

### DISCUSSION THREAD

**Discussion by ghadzhigeorgiev** [dotnet/csharplang#7718] [Language Ideas]
Will be nice to allow something like
```
public class User (string Name) <= (ILogger<User> logger) : Person, IEmployee
{...}
```
or
```
public class User (string Name) : Person, IEmployee <= (ILogger<User> logger)
{...}
```
It will be just a syntactic sugar but will be easy on the eye and will reflect better the intention.

This should not prevent one from using also
```
public class User (string Name, IConfiguration config) <= (ILogger<User> logger) : Person, IEmployee
{...} 
```
or old way
```
public class User (string Name, IConfiguration config, ILogger<User> logger) : Person, IEmployee
{...} 
```
which will keep existing code compatible.

**Comment by HaloFour** [comment]
Why does this need language support?  You can already very easily manage DI via constructor, properties or other forms of injection through the supported features of whatever IoC container you are using.

**Reply by jnm2** [reply]
And with required properties or primary constructors, DI just got a lot nicer.

**Comment by ghadzhigeorgiev** [comment]
It's cleaner, for example:
```
public class User (string Name, string Email, DateTime BirthDate) : Person, IEmployee <= (ILogger<User> logger, Ixxxx<yyy> zzz, etc... )
{...}
```

One can clearly see what's part of the class, what's inherited and what are the interfaces, and lastly, what class needs to be injected from outside.

Now compare to
```
public class User (string Name, string Email, DateTime BirthDate, ILogger<User> logger, Ixxxx<yyy> zzz, etc... ): Person, IEmployee
{...}
```
quickly becomes messy.

Behind the scenes, compiler will generate same code and also, if one prefers the other ways, they still be there.

**Reply by HaloFour** [reply]
> One can clearly see what's part of the class, what's inherited and what are the interfaces, and lastly, what class needs to be injected from outside.

I disagree.  I think the proposed syntax is messier as it mixes the concerns around.  You don't know if those are constructor parameters, properties, both, whatever.  You don't see what it's doing.  Worse, it makes DI a compiler concern, whereas it really shouldn't be.

> Behind the scenes, compiler will generate same code and also, if one prefers the other ways, they still be there.

If you're using DI with an IoC container that's something a user would never see anyway.

I really dislike the notion of making DI/IoC a language concern.  And your example makes no sense in the context of DI given that it's a DTO, and should have no dependencies.  That's made extra awkward by trying to mix data properties like a name or an email address.

**Reply by ghadzhigeorgiev** [reply]
How moving out ILogger parameter out of primary constructor parameters is different to compiler? It will generate same code, it is just placed in different place in the primary constructor syntax.

**Reply by HaloFour** [reply]
It's not placed in the primary constructor syntax, that's (part of) the problem.  Ultimately, the compiler shouldn't know nor care where those parameters come from, or that they have anything to do with DI.  If they are satisfied via an IoC container, wonderful, but that's not for the compiler to care about.  And honestly this use case doesn't make sense.  If `User` is a component with dependencies, why does it declare a constructor for parameters that _aren't_ dependencies?  Where do they come from?  If the user is creating these things directly, then something is already wrong.  `User` is either a data class or a component, it shouldn't be both.

**Reply by ghadzhigeorgiev** [reply]
User is just for illustration, do not focus on minute detail why User, could be Person, Car, BankAccount, whatever... this is not the point.

**Reply by HaloFour** [reply]
It is an important point, though.  If you have a type that has a combination of data elements and dependencies, something is wrong.  There shouldn't be a need for a feature like this because either every parameter is a dependency, or none of them are.

**Reply by ghadzhigeorgiev** [reply]
Again, we are not talking domain design....

**Reply by HaloFour** [reply]
You're asking for a language feature to satisfy a use case with dependency injection.  Language features are ridiculously expensive.  The feature has to demonstrate how it satisfies general purpose use cases.  If it's extremely niche because it doesn't even fit the normal parameters of dependency injection, then it has less purpose.  I will argue that declaring a type where some parameters are dependencies and some are data are not typical in dependency injection, so this syntax doesn't solve use cases.

**Reply by ghadzhigeorgiev** [reply]
I may said that incorrectly, but after all it comes to Syntactic sugar, as title says. C# is now full of them.

**Reply by HaloFour** [reply]
Not all syntax sugar is equal.  Splitting up the parameters of a primary constructor doesn't make sense to me.  Having a syntax which makes it appear as if a member has one signature when it really has a different signature is not appealing to me.

**Reply by ghadzhigeorgiev** [reply]
And that's cool. I can accept that as opinion and personal preference, that's why I'm expecting this not to break other ways.
I myself, still prefer to see immediately "what's part of the class, what's inherited and what are the interfaces, and lastly, what class needs to be injected from outside".

**Reply by HaloFour** [reply]
> and lastly, what class needs to be injected from outside

IMO, the language already provides a way to declare that: constructor parameters and required properties
